README
NexStarCtl version 0.15
=======================

This module implements the serial commands supported by the Celestron
NexStar hand control (HC) and SkyWatcher/Orion SynScan HC. This applies
to the Celestron: NexStar GPS, NexStar GPS-SA, NexStar iSeries, NexStar
SE Series, NexStar GT, CPC, SLT, Advanced-VX, Advanced-GT, CGE, CGEM etc.
SkyWatcher: EQ5, HEQ5, EQ6 (Pro), AZ-EQ5 GT, AZ-EQ6 GT, EQ8 etc. and the
corresponding Orion mounts.

Communication to the hand control is 9600 bits/sec, no parity and one 
stop bit via the RS-232 port on the base of the hand control.
Communication can be established over TCP/IP if nexbridge is running
on the computer connected to the telescope.

For extended example how to use this perl module look in to the
distribution folder for  nexstarctl/nexstarctl.pl. This program is a
complete console tool to control NexStar telescopes based on NexStarCtl
module.

NOTE: For SkyWatcher/Orion mounts it is highly recommended to enforce
protocol version checking (see enforce_proto_version()) as the AUX
commands are not supported or may behave erratically.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

   Device::SerialPort
   Time::Local

COPYRIGHT AND LICENCE

Copyright (C) 2013 by Rumen G.Bogdanovski

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension NexStarCtl.
	- add support for nexstar:// prefix
0.15  Thu Apr  7 15:21:03 2016
	- add several new calls
	- add support for SkyWatcher mounts
	- add support for different protocol versions
	- bug/typo fixes

0.14  Fri Sep  4 16:24:41 2015
	- fix tc_set_location()

0.13  Fri Sep 19 15:04:10 2014
	- added TCP/IP telescope control

0.11  Wed Mar 12 11:28:25 2014
	- ported nexstarctl.pl to windows OS.
	- several bugfixes
	- CGEM mount is now recognized
	- decreased required perl version

      Mon Mar 10 01:25:12 2014
	- initial port to Windows OS.
	- fix several typos

0.10  Sun Mar  9 21:20:07 2014
	- added aux commands tc_get_autoguide_rate
	  tc_set_autoguide_rate tc_get_backlash tc_set_backlash

0.09  Sat Dec  7 22:21:45 2013
	- created nexstarctl.pl tool
	- small fixes to NexStarCtl.pm

0.08  Wed Dec  4 21:36:19 2013
	- tc_set_time() fix misstype in daylight saving logic

      Tue Dec  3 14:38:36 2013
	- tc_set_time() sets date/time to RTC too, if available

0.06.1 Tue Dec  3 12:24:58 2013
	- fix some documentation errors

0.06  Sun Nov  3 13:16:30 2013
	- fix the ID of Advanced VX

0.05  Fri Oct 25 17:01:44 2013
	- add get_model_name()

0.04  Sun Sep 29 15:40:22 2013
	- Fix read_telescope() it needs $len parameter because '#' does not 
	  always indicate end of responce.
	- changed port data type to raw.

0.03  Sat Sep 21 22:06:39 2013
	- Fix Declination -90deg boder condition

0.02  Fri Aug 30 00:33:20 2013
	- Fix message lenght in variable rate slew command

0.01  Thu Aug 29 12:04:24 2013
	- initial version


Makefile.PL
use 5.012004;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

my $required;
if ($^O eq "MSWin32") {
	$required = { 'Win32::SerialPort' => 0.1 };
} else {
	$required = { 'Device::SerialPort' => 0.1 };
}

WriteMakefile(
    NAME              => 'NexStarCtl',
    VERSION_FROM      => 'lib/NexStarCtl.pm', # finds $VERSION
    PM                => { 'nexstarctl/nexstarctl.pl' => '$(INST_BIN)/nexstarctl',
                           'lib/NexStarCtl.pm' => '$(INST_LIB)/NexStarCtl.pm'
                         },
    BUILD_REQUIRES    => $required,
    PREREQ_PM         => $required,
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT  => 'NexStarCtl - API to control NexStar compatible telescopes',
       AUTHOR         => 'Rumen Bogdanovski <rumen@skyarchive.org>') : ()),
);


nexstarctl/nexstarctl.pl
#!/usr/bin/perl
###############################################################
# Celestron NexStar telescope control tool. Its primary purpose
# is to illustrate how NexStarCtl module can be used, but it is
# a useful tool for telescope automation.
#
# This is a GPL software, created by Rumen G. Bogdanovski.
#                                              December 2013
###############################################################

use strict;
use NexStarCtl;
use Date::Parse;
use Time::Local;
use File::Basename;
use POSIX qw(strftime);
use Getopt::Std;
use if $^O eq "MSWin32", "Win32::Console::ANSI";
use Term::ANSIColor qw(:constants);
$Term::ANSIColor::AUTORESET = 1;

my $VERSION = "0.5";

my $port;
my $verbose;

sub print_help() {
	my $N = basename($0);
	print "\n".
	      "Celestron NexStar telescope control tool v.$VERSION. Its primary purpose is to illustrate\n".
	      "how NexStarCtl module can be used, but it is a useful tool for telescope automation.\n",
	      "This is a GPL software, created by Rumen G. Bogdanovski.\n".
	      "\n".
	      "Usage: $N info [telescope]\n".
	      "       $N settime \"DATE TIME\" TZ ISDST [telescope]\n".
	      "       $N settime [telescope]\n".
	      "       $N gettime [telescope]\n".
	      "       $N setlocation LON LAT [telescope]\n".
	      "       $N getlocation [telescope]\n".
	      "       $N settrack [equatorial|north|south|azalt|off] [telescope]\n".
	      "       $N gettrack [telescope]\n".
	      "       $N goto RA DE [telescope]\n".
	      "       $N gotoaz AZ ALT [telescope]\n".
	      "       $N sync RA DE [telescope]\n".
	      "       $N getrade [telescope]\n".
	      "       $N getazalt [teelescope]\n".
	      "       $N abort [telescope]\n".
	      "       $N status [telescope]\n".
	      "       $N netscan\n".
	      "options:\n".
	      "       -v verbose output\n".
	      "       -h print this help\n".
	      "[telescope]:\n".
	      "       The telescope port could be specified with this parameter or TELESCOPE_PORT\n".
	      "       environment variable can be set. Defaults depend on the operating system:\n".
	      "          Linux: /dev/ttyUSB0\n".
	      "          MacOSX: /dev/cu.usbserial\n".
	      "          Solaris: /dev/ttya\n".
	      "          Windows: COM1\n".
	      "       The network connected telescopes can be listed with \"$N netscan\"\n".
	      "       and the value of TELESCOPE_PORT for each telescope is given.\n\n".
	      "NOTE: RA and DE should be secified in J2000 and are returned in J2000.\n\n";
}


sub check_align($) {
	my ($dev) = @_;
	my $align = tc_check_align($dev);
	if (!defined $align) {
		print RED "Error reading from telescope.\n";
		return undef;
	} elsif ($align == 0) {
		print RED "The telescope is NOT aligned. Please complete the alignment routine.\n";
		return undef;
	}
	$verbose && print GREEN "The telescope is aligned.\n";
}


sub init_telescope {
	my ($tport, $check) = @_;
	my $dev = open_telescope_port($tport);
	if (!defined($dev)) {
		print RED "Can\'t open telescope on port $tport: $!\n";
		return undef;
	}

	my $res = enforce_protocol_version($dev);
	if ($res == undef) {
		print RED "Communication with the telescope on port $tport failed: $!\n";
		close_telescope_port($dev);
		return undef;
	}

	$verbose && print GREEN "The telescope port $tport is open.\n";

	if (! $check) {
		return $dev;
	}

	# Check the telescope is slewing return failure
	if (tc_goto_in_progress($dev)) {
		print RED "GOTO in progress, try again.\n";
		close_telescope_port($dev);
		return undef;
	}

	return $dev;
}


sub info {
	my @params = @_;
	if ($#params <= 0) {
		if (defined $params[0]) {
			$port = $params[0];
		}
	} else {
		print RED "info: Wrong parameters.\n";
		return undef;
	}

	my $dev = init_telescope($port);
	return undef if (! defined $dev);

	print "Driver: NexStarCtl v.$NexStarCtl::VERSION\n";

	my $echo = tc_echo($dev,"X");
	if ($echo ne "X") {
		print RED "info: No telescope mount found on port $port\n";
		close_telescope_port($dev);
		return undef;
	}
	print "Mount port: $port\n";

	my $model = tc_get_model($dev);
	if (!defined $model) {
		print RED "info: Error getting model. $!\n";
		close_telescope_port($dev);
		return undef;
	}
	my $model_name = get_model_name($model);
	print "Mount model: $model_name ($model)\n";

	my $version = tc_get_version($dev);
	if (!defined $version) {
		print RED "info: Error getting version. $!\n";
		close_telescope_port($dev);
		return undef;
	}
	print "HC version: $version\n";

	close_telescope_port($dev);
	return 1;
}


sub status {
	my @params = @_;
	if ($#params <= 0) {
		if (defined $params[0]) {
			$port = $params[0];
		}
	} else {
		print RED "status: Wrong parameters.\n";
		return undef;
	}

	my $dev = init_telescope($port);
	return undef if (! defined $dev);

	my $status = tc_goto_in_progress($dev);
	if (!defined $status) {
		print RED "status: Error getting goto status. $!\n";
		close_telescope_port($dev);
		return undef;
	}

	my $tracking = tc_get_tracking_mode($dev);
	if (!defined $tracking) {
		if ($NexStarCtl::error != -5) {
			print RED "status: Error geting tracking mode. $!\n";
			close_telescope_port($dev);
			return undef;
		} else {  # get tracking mode is unsupported on this mount
			if ($status == 0) {
				print "Telescope is either tracking or not, but GOTO is not in progress.\n";
			} else {
				print "GOTO is in progress.\n";
			}
		}
	} else {
		if (($status == 0) && ($tracking == TC_TRACK_OFF)) {
			print "Telescope is not tracking.\n";
		} elsif (($status == 0) && ($tracking != TC_TRACK_OFF)) {
			print "Telescope is tracking.\n";
		} else {
			print "GOTO is in progress.\n";
		}
	}
	close_telescope_port($dev);
	return 1;
}


sub gettime {
	my @params = @_;
	if ($#params <= 0) {
		if (defined $params[0]) {
			$port = $params[0];
		}
	} else {
		print RED "gettime: Wrong parameters.\n";
		return undef;
	}

	my $dev = init_telescope($port);
	return undef if (! defined $dev);

	my ($date, $time, $tz, $isdst) = tc_get_time_str($dev);
	if (! defined $date) {
		print RED "gettime: Failed. $!\n";
		close_telescope_port($dev);
		return undef;
	}
	print "$date $time, TZ = $tz, DST = $isdst\n";

	close_telescope_port($dev);
	return 1;
}


sub settime {
	my @params = @_;
	my $date;
	my $tz;
	my $time;
	my $isdst;

	if ($#params == 2) {
		$date = $params[0];
		$tz = round($params[1]);
		$isdst = $params[2];

	} elsif ($#params == 3) {
		$date = $params[0];
		$tz = round($params[1]);
		$isdst = $params[2];
		$port = $params[3];

	} elsif ($#params <= 0) {
		if (defined $params[0]) {
			$port = $params[0];
		}
		$time=time();
		$isdst = (localtime($time))[-1];
		$tz = int((timegm(localtime($time)) - $time) / 3600);
		$tz = $tz-1 if ($isdst);

	} else {
		print RED "settime: Wrong parameters.\n";
		return undef;
	}

	if (($tz < -12) or ($tz > 12)) {
		print RED "settime: Wrong time zone.\n";
		return undef;
	}

	# if $date is defined => the date is given by user
	if (defined $date) {
		$time = str2time($date);
		if (!defined $time) {
			print RED "settime: Wrong date format.\n";
			return undef;
		}
	}

	# Do not set the time if the telescope is slewing
	# so set second parameter to 1
	my $dev = init_telescope($port, 1);
	return undef if (! defined $dev);

	my ($s, $m, $h, $day, $mon, $year) = localtime($time);
	my $time_str = sprintf("%2d:%02d:%02d", $h, $m, $s);
	my $date_str = sprintf("%02d-%02d-%04d", $day, $mon + 1, $year + 1900);
	$verbose && print "settime: $date_str $time_str, TZ = $tz, DST = $isdst\n";

	if (! tc_set_time($dev, $time, $tz, $isdst)) {
		print RED "settime: Failed. $!\n";
		close_telescope_port($dev);
		return undef;
	}

	close_telescope_port($dev);
	return 1;
}


sub getlocation {
	my @params = @_;
	if ($#params <= 0) {
		if (defined $params[0]) {
			$port = $params[0];
		}
	} else {
		print RED "getlocation: Wrong parameters.\n";
		return undef;
	}

	my $dev = init_telescope($port);
	return undef if (! defined $dev);

	my ($lon,$lat) = tc_get_location_str($dev);
	if (! defined $lon) {
		print RED "getlocation: Failed. $!\n";
		close_telescope_port($dev);
		return undef;
	}
	print "$lon, $lat\n";

	close_telescope_port($dev);
	return 1;
}


sub setlocation {
	my @params = @_;
	my $lon;
	my $lat;

	if ($#params == 1) {
		$lon = $params[0];
		$lat = $params[1];

	} elsif ($#params == 2) {
		$lon = $params[0];
		$lat = $params[1];
		$port = $params[2];

	} else {
		print RED "settime: Wrong parameters.\n";
		return undef;
	}

	# Convert strings $lon/$lat to decimal $lond/latd in decimal degrees
	my $lond = dms2d($lon);
	if ((!defined $lond) or ($lond > 180) or ($lond < -180)) {
		print RED "setlocation: Wrong longitude.\n";
		return undef;
	}

	my $latd = dms2d($lat);
	if ((!defined $latd) or ($latd > 180) or ($latd < -180)) {
		print RED "setlocation: Wrong latitude.\n";
		return undef;
	}

	# do not set location if the telescope is slewing
	# so set second parameter to 1
	my $dev = init_telescope($port, 1);
	return undef if (! defined $dev);

	$verbose && print "setlocation: lon = $lond, lat = $latd\n";

	if (! tc_set_location($dev, $lond, $latd)) {
		print RED "setlocation: Failed. $!\n";
		close_telescope_port($dev);
		return undef;
	}

	close_telescope_port($dev);
	return 1;
}


sub settrack {
	my @params = @_;
	my $mode;

	if ($#params == 0) {
		$mode = $params[0];

	} elsif ($#params == 1) {
		$mode = $params[0];
		$port = $params[1];

	} else {
		print RED "settrack: Wrong parameters.\n";
		return undef;
	}

	my $tracking;
	if ($mode eq "equatorial") {
		$tracking = TC_TRACK_EQ;
	} elsif ($mode eq "south") {
		$tracking = TC_TRACK_EQ_SOUTH;
	} elsif ($mode eq "north") {
		$tracking = TC_TRACK_EQ_NORTH;
	} elsif ($mode eq "azalt") {
		$tracking = TC_TRACK_ALT_AZ;
	} elsif ($mode eq "off") {
		$tracking = TC_TRACK_OFF;
	} else {
		print RED "settrack: Wrong parameters.\n";
		return undef;
	}

	# do not set tracking mode if slewing
	# so set second parameter to 1
	my $dev = init_telescope($port, 1);
	return undef if (! defined $dev);

	if (! tc_set_tracking_mode($dev, $tracking)) {
		print RED "settrack: Error setting tracking mode. $!\n";
		close_telescope_port($dev);
		return undef;
	}

	$verbose && print "settrack: $mode ($tracking)\n";

	close_telescope_port($dev);
	return 1;
}


sub gettrack {
	my @params = @_;
	if ($#params <= 0) {
		if (defined $params[0]) {
			$port = $params[0];
		}
	} else {
		print RED "gettrack: Wrong parameters.\n";
		return undef;
	}

	my $dev = init_telescope($port);
	return undef if (! defined $dev);

	my $tracking = tc_get_tracking_mode($dev);
	if (!defined $tracking) {
		print RED "gettrack: Error geting tracking mode. $!\n";
		close_telescope_port($dev);
		return undef;
	}

	if ($tracking == TC_TRACK_OFF) {
		print "Tracking: OFF\n";
	} elsif ($tracking == TC_TRACK_EQ_SOUTH) {
		print "Tracking: Equatorial South\n";
	} elsif ($tracking == TC_TRACK_EQ_NORTH) {
		print "Tracking: Equatorial North\n";
	} elsif ($tracking == TC_TRACK_ALT_AZ) {
		print "Tracking: Aaltazimuthal\n";
	} elsif ($tracking == TC_TRACK_EQ) {
		print "Tracking: Equatorial\n";
	} elsif ($tracking == TC_TRACK_EQ_PEC) {
		print "Tracking: Equatorial + PEC\n";
	} else {
		print "Tracking: Unknown\n";
	}

	close_telescope_port($dev);
	return 1;
}


sub gotoeq {
	my @params = @_;
	my $ra;
	my $de;

	if ($#params == 1) {
		$ra = $params[0];
		$de = $params[1];

	} elsif ($#params == 2) {
		$ra = $params[0];
		$de = $params[1];
		$port = $params[2];

	} else {
		print RED "goto: Wrong parameters.\n";
		return undef;
	}

	# convert strings $ra/$de to $rad/$ded in decimal degrees
	my $rad = hms2d($ra);
	if ((!defined $rad) or ($rad > 360) or ($rad < 0)) {
		print RED "goto: Wrong rightascension.\n";
		return undef;
	}

	my $ded = dms2d($de);
	if ((!defined $ded) or ($ded > 90) or ($ded < -90)) {
		print RED "goto: Wrong declination.\n";
		return undef;
	}

	# if telescope is slewing do nothing
	# so set second parameter to 1
	my $dev = init_telescope($port, 1);
	return undef if (! defined $dev);

	my $align = tc_check_align($dev);
	if (!defined $align) {
		print RED "goto: Error reading from telescope.\n";
		return undef;
	} elsif ($align == 0) {
		print RED "The telescope is NOT aligned. Please align before GOTO.\n";
		return undef;
	}
	$verbose && print GREEN "The telescope is aligned.\n";

	$verbose && print "goto: ra = $rad, dec = $ded\n";

	if (! tc_goto_rade_p($dev, $rad, $ded)) {
		print RED "goto: Failed. $!\n";
		close_telescope_port($dev);
		return undef;
	}

	print "GOTO started...\n";

	close_telescope_port($dev);
	return 1;
}

sub sync {
	my @params = @_;
	my $ra;
	my $de;

	if ($#params == 1) {
		$ra = $params[0];
		$de = $params[1];

	} elsif ($#params == 2) {
		$ra = $params[0];
		$de = $params[1];
		$port = $params[2];

	} else {
		print RED "sync: Wrong parameters.\n";
		return undef;
	}

	# convert strings $ra/$de to $rad/$ded in decimal degrees
	my $rad = hms2d($ra);
	if ((!defined $rad) or ($rad > 360) or ($rad < 0)) {
		print RED "sync: Wrong rightascension.\n";
		return undef;
	}

	my $ded = dms2d($de);
	if ((!defined $ded) or ($ded > 90) or ($ded < -90)) {
		print RED "sync: Wrong declination.\n";
		return undef;
	}

	# if telescope is slewing do nothing
	# so set second parameter to 1
	my $dev = init_telescope($port, 1);
	return undef if (! defined $dev);

	my $align = tc_check_align($dev);
	if (!defined $align) {
		print RED "sync: Error reading from telescope.\n";
		return undef;
	} elsif ($align == 0) {
		print RED "The telescope is NOT aligned. Please align before sync.\n";
		return undef;
	}
	$verbose && print GREEN "The telescope is aligned.\n";

	$verbose && print "sync: ra = $rad, dec = $ded\n";

	if (! tc_sync_rade_p($dev, $rad, $ded)) {
		print RED "sync: Failed. $!\n";
		close_telescope_port($dev);
		return undef;
	}

	close_telescope_port($dev);
	return 1;
}


sub getrade {
	my @params = @_;
	if ($#params <= 0) {
		if (defined $params[0]) {
			$port = $params[0];
		}
	} else {
		print RED "getrade: Wrong parameters.\n";
		return undef;
	}

	my $dev = init_telescope($port);
	return undef if (! defined $dev);

	my ($ra,$de) = tc_get_rade_p($dev);
	if (!defined $ra) {
		print RED "getrade: Error geting ra/de. $!\n";
		close_telescope_port($dev);
		return undef;
	}

	# convert digital degrees $ra/$de to strings and print them
	print d2hms($ra) . ", " . d2dms($de)."\n";

	close_telescope_port($dev);
	return 1;
}


sub gotoaz {
	my @params = @_;
	my $az;
	my $alt;

	if ($#params == 1) {
		$az = $params[0];
		$alt = $params[1];

	} elsif ($#params == 2) {
		$az = $params[0];
		$alt = $params[1];
		$port = $params[2];

	} else {
		print RED "gotoaz: Wrong parameters.\n";
		return undef;
	}

	# convert strings $az/$alt to $az/$alt in decimal degrees
	my $azd = dms2d($az);
	if (!defined $azd) {
		print RED "gotoaz: Wrong azimuth.\n";
		return undef;
	}

	my $altd = dms2d($alt);
	if ((!defined $altd) or ($altd > 90) or ($altd < -90)) {
		print RED "gotoaz: Wrong altitude.\n";
		return undef;
	}

	# if telescope is slewing do nothing
	# so set second parameter to 1
	my $dev = init_telescope($port, 1);
	return undef if (! defined $dev);

	my $align = tc_check_align($dev);
	if (!defined $align) {
		print RED "gotoaz: Error reading from telescope.\n";
		return undef;
	} elsif ($align == 0) {
		print RED "The telescope is NOT aligned. Please align before GOTO.\n";
		return undef;
	}
	$verbose && print GREEN "The telescope is aligned.\n";

	$verbose && print "gotoaz: az = $azd, alt = $altd\n";

	if (! tc_goto_azalt_p($dev, $azd, $altd)) {
		print RED "gotoaz: Failed. $!\n";
		close_telescope_port($dev);
		return undef;
	}

	print "GOTO started...\n";

	close_telescope_port($dev);
	return 1;
}


sub getazalt {
	my @params = @_;
	if ($#params <= 0) {
		if (defined $params[0]) {
			$port = $params[0];
		}
	} else {
		print RED "getazalt: Wrong parameters.\n";
		return undef;
	}

	my $dev = init_telescope($port);
	return undef if (! defined $dev);

	my ($az,$alt) = tc_get_azalt_p($dev);
	if (!defined $az) {
		print RED "getazalt: Error geting az/alt. $!\n";
		close_telescope_port($dev);
		return undef;
	}

	print d2dms($az) . ", " . d2dms($alt)."\n";

	close_telescope_port($dev);
	return 1;
}


sub abort {
	my @params = @_;
	if ($#params <= 0) {
		if (defined $params[0]) {
			$port = $params[0];
		}
	} else {
		print RED "abort: Wrong parameters.\n";
		return undef;
	}

	my $dev = init_telescope($port);
	return undef if (! defined $dev);

	my $reult = tc_goto_cancel($dev);
	if (!defined $reult) {
		print RED "abort: Failed. $!\n";
		close_telescope_port($dev);
		return undef;
	}

	$verbose && print "GOTO aborted.\n";

	close_telescope_port($dev);
	return 1;
}


sub netscan {
	eval "use Net::Bonjour";
	if ($@) {
		print RED "This feature is disabled to enable it Net::Bonjour should be installed!\n";
		return undef;
	}
	my $res = Net::Bonjour->new('nexbridge');

	$res->discover();

	if ( $res->entries == 0) {
		print "\nNo telescopes discovered!\n";
		print YELLOW "To discover the network connected telescopes they should be\n";
		print YELLOW "exported via Bonjour with nexbridge (See perldoc NexStarCtl)\n\n";
		return 1;
	}

	print "\n";
	printf "%-30s %s\n", "NAME", "TELESCOPE_PORT";
	printf "%-30s %s\n", "===============", "==============";

	$res->discover();
	foreach my $entry ($res->entries) {
		printf "%-30s %s://%s:%s (%s)\n", $entry->name, $res->protocol, $entry->hostname, $entry->port, $entry->address;
	}
	print "\n";

        return 1;
}


sub main() {
	my %options = ();

	my $command = shift @ARGV;

	if (defined $ENV{TELESCOPE_PORT}) {
		$port = $ENV{TELESCOPE_PORT};
	} else {
		if ($^O eq 'linux') {
			$port = "/dev/ttyUSB0";
		} elsif ($^O eq 'darwin') {
			$port = "/dev/cu.usbserial";
		} elsif ($^O eq 'solaris') {
			$port = "/dev/ttya";
		} elsif ($^O eq 'MSWin32') {
			$port = "COM1";
		}
	}

	if (getopts("vh", \%options) == undef) {
		exit 1;
	}

	if (defined($options{h}) or (!defined($command))) {
		print_help();
		exit 1;
	}

	if(defined($options{v})) {
		$verbose = 1;
	}

	if ($command eq "info") {
		if (! info(@ARGV)) {
			$verbose && print RED "Get info returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Get info succeeded.\n";
		exit 0;

	} elsif ($command eq "gettime") {
		if (! gettime(@ARGV)) {
			$verbose && print RED "Get time returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Get time succeeded.\n";
		exit 0;

	} elsif ($command eq "settime") {
		if (! settime(@ARGV)) {
			$verbose && print RED "Set time returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Set time succeeded.\n";
		exit 0;

	} elsif ($command eq "getlocation") {
		if (! getlocation(@ARGV)) {
			$verbose && print RED "Get location returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Get location succeeded.\n";
		exit 0;

	} elsif ($command eq "setlocation") {
		if (! setlocation(@ARGV)) {
			$verbose && print RED "Set location returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Set location succeeded.\n";
		exit 0;

	} elsif ($command eq "settrack") {
		if (! settrack(@ARGV)) {
			$verbose && print RED "Set track returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Set track succeeded.\n";
		exit 0;

	} elsif ($command eq "gettrack") {
		if (! gettrack(@ARGV)) {
			$verbose && print RED "Get track returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Get track succeeded.\n";
		exit 0;

	} elsif ($command eq "goto") {
		if (! gotoeq(@ARGV)) {
			$verbose && print RED "GOTO ra/de returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "GOTO ra/de succeeded.\n";
		exit 0;

	} elsif ($command eq "sync") {
		if (! sync(@ARGV)) {
			$verbose && print RED "SYNC ra/de returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "SYNC ra/de succeeded.\n";
		exit 0;

	} elsif ($command eq "gotoaz") {
		if (! gotoaz(@ARGV)) {
			$verbose && print RED "GOTO az/alt returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "GOTO az/alt succeeded.\n";
		exit 0;

	} elsif ($command eq "getrade") {
		if (! getrade(@ARGV)) {
			$verbose && print RED "Get ra/de returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Get ra/de succeeded.\n";
		exit 0;

	} elsif ($command eq "getazalt") {
		if (! getazalt(@ARGV)) {
			$verbose && print RED "Get az/alt returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Get az/alt succeeded.\n";
		exit 0;

	} elsif ($command eq "abort") {
		if (! abort(@ARGV)) {
			$verbose && print RED "Abort returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Abort succeeded.\n";
		exit 0;

	} elsif ($command eq "status") {
		if (! status(@ARGV)) {
			$verbose && print RED "Status returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Status succeeded.\n";
		exit 0;

	} elsif ($command eq "netscan") {
		if (! netscan()) {
			$verbose && print RED "Netscan returned error.\n";
			exit 1;
		}
		$verbose && print GREEN "Netscan succeeded.\n";
		exit 0;

	} elsif ($command eq "-h") {
		print_help();
		exit 0;

	} else {
		print RED "There is no such command \"$command\".\n";
	}
}

main;


testcode/celestron.pl
#!/usr/bin/perl
use strict;
use lib '../lib';
use NexStarCtl;


# print "(0,".TC_AXIS_RA_AZM.",".TC_DIR_POSITIVE.",3)\n";

#enforce_vendor_protocol(VNDR_CELESTRON);

#print get_mcdel_name(1)."\n";
#print get_model_name(2)."\n";
#print get_model_name(22)."\n"; 

my $port = open_telescope_port("nexstar://localhost:2000"); 
#my $port = open_telescope_port("/tmp/cam");

if (!defined $port) {
	print "Can not open communication port.\n";
	exit;
}

enforce_vendor_protocol(VNDR_CELESTRON);
enforce_protocol_version($port,0x040001);

print "tc_get_orientation = ".tc_get_orientation($port)."    err:".$NexStarCtl::error."\n";
print "tc_slew_fixed = ".tc_slew_fixed($port,1,1)."    err:".$NexStarCtl::error."\n";

#exit(0); 


print "MOUNT = ".tc_get_model($port)."\n";

print "VERSION = ".tc_get_version($port)."\n";

#enforce_protocol_version($port,VER_1_2);
#enforce_vendor_protocol(VNDR_CELESTRON);

print "Align = ".tc_check_align($port)."\n";

print "orientation = ".tc_get_orientation($port)."\n";
print "Error=".$NexStarCtl::error."\n";

#my $response = tc_pass_through_cmd($port, 1, 178, 4, 0, 0, 0, 2);
#print "GPS is present:". ord($response). "\n";

#print "SET LOC= ".tc_set_location($port,  dms2d("22:58:09"), dms2d("44:05:31"))."\n";

#my ($lon,$lat) = tc_get_location($port);
#print "LON=$lon LAT=$lat\n";

#my ($lon,$lat) = tc_get_location_str($port);
#print "LONs=$lon LATs=$lat\n";

#my $tm=time();
#print "$tm\n";
#print "SETTIME= ".tc_set_time($port,$tm,2,0)."\n";

#my ($date,$time,$tz,$dst) = tc_get_time_str($port);

#print "$date, $time, $tz, $dst\n";


#my $echo;
#for (my $i=1; $i<256; $i++) {
#	$echo = ord(tc_echo($port, chr($i)));
#
#	#sleep 1;
#	if ($echo != $i) {
#		print "ERROR: Sent $i received $echo\n";
#	} else {
#		print "OK: Sent $i received $echo\n";
#	}
#	
#}

print "GOTO = ".tc_goto_rade_p($port,11*15,21)."\n";

sleep(2);
print "GOTO Cancel = ".tc_goto_cancel($port)."\n";

print "TRACKING SET= ".tc_set_tracking_mode($port,0)."\n";
print "TRACKING = ".tc_get_tracking_mode($port)."\n";

tc_slew_variable($port,TC_AXIS_RA_AZM,TC_DIR_POSITIVE,15.25); 
sleep(2);
tc_slew_variable($port,TC_AXIS_DE_ALT,TC_DIR_POSITIVE,1);
sleep(2);
tc_slew_fixed($port,TC_AXIS_RA_AZM,TC_DIR_NEGATIVE,0);
sleep(2);
tc_slew_fixed($port,TC_AXIS_DE_ALT,TC_DIR_POSITIVE,0);

print "TRACKING = ".tc_get_tracking_mode($port)."\n";

print "TRACKING SET= ".tc_set_tracking_mode($port,2)."\n";
print "TRACKING = ".tc_get_tracking_mode($port)."\n";


#while (tc_goto_in_progress($port)) {
#	sleep(1);
#	print ".";
#}
#print "OK\n";

#print "SYNC = ".tc_sync_rade_p($port,0,0)."\n";

my ($rap,$decp) = tc_get_rade_p($port);
my ($ra,$dec) = tc_get_rade($port);


my ($azp,$altp) = tc_get_azalt_p($port);
my ($az,$alt) = tc_get_azalt($port);

#print "Set RA(+) Backlash: " . tc_set_backlash($port,TC_AXIS_RA_AZM,TC_DIR_POSITIVE,0) . "\n";
#print "Set RA(-) Backlash: " . tc_set_backlash($port,TC_AXIS_RA_AZM,TC_DIR_NEGATIVE,0) . "\n";
#print "Set DE(+) Backlash: " . tc_set_backlash($port,TC_AXIS_DE_ALT,TC_DIR_POSITIVE,0) . "\n";
#print "Set DE(-) Backlash: " . tc_set_backlash($port,TC_AXIS_DE_ALT,TC_DIR_NEGATIVE,0) . "\n";
#print "Get RA(+) Backlash: " . tc_get_backlash($port,TC_AXIS_RA_AZM,TC_DIR_POSITIVE) . "\n";
#print "Get RA(-) Backlash: " . tc_get_backlash($port,TC_AXIS_RA_AZM,TC_DIR_NEGATIVE) . "\n";
#print "Get DE(+) Backlash: " . tc_get_backlash($port,TC_AXIS_DE_ALT,TC_DIR_POSITIVE) . "\n";
#print "Get DE(-) Backlash: " . tc_get_backlash($port,TC_AXIS_DE_ALT,TC_DIR_NEGATIVE) . "\n";

print "Get RA autoguide rate: " . tc_get_autoguide_rate($port,TC_AXIS_RA_AZM) . "\n";
print "Get DE autoguide rate: " . tc_get_autoguide_rate($port,TC_AXIS_DE_ALT) . "\n";

for (my $rate = -1; $rate < 102; $rate++) {
	my $res = tc_set_autoguide_rate($port,TC_AXIS_RA_AZM,$rate);
	if ($res != 1) {print "set faied\n";}
	print "Get RA autoguide rate: $rate => " . tc_get_autoguide_rate($port,TC_AXIS_RA_AZM) . "\n";
	#print "\n";
	#print "Set DE autoguide rate: $rate => " . tc_set_autoguide_rate($port,TC_AXIS_DE_ALT,22) . "\n";
	#print "Get DE autoguide rate: " . tc_get_autoguide_rate($port,TC_AXIS_DE_ALT) . "\n";
}

close_telescope_port($port);


#my $ra1=hms2d("22:33:32");
#my $dec1=dms2d("-44:22:11");

#my $ra1=222.333;
#my $dec1=-88.9999;

#my $nex="1000,1000";
#my ($ra1,$dec1)=nex2dd($nex);

#my ($ra2,$dec2)=precess($ra1,$dec1,2013.6,2013);

#my $nex=dd2nex($ra1,$dec1);
#print "NEX1=$nex\n";

#my $nex2=dd2nex($ra2,$dec2);
#print "NEX2=$nex2\n";

#my ($ra,$dec)=nex2dd($nex);
#print "RA=$ra\nDE=$dec\n";

#my $ras=d2hms($ra);
#my $decs=d2dms($dec);

#my $rasp=d2hms($rap);
#my $decsp=d2dms($decp);


#print "RA=$ras\nDEC=$decs\n";
#print "RAp=$rasp\nDECp=$decsp\n";

#print "AZ=$az\nALT=$alt\n";
#print "AZp=$azp\nALTp=$altp\n";
